### ISSUE TITLE
Why is `VolumeCapability` a list in `CreateVolumeRequest`?

### ISSUE THREAD

**Issue by saad-ali** [container-storage-interface/spec#250]
@msau42 was asking me why `VolumeCapability` is a repeated field in `CreateVolumeRequest` and some other places:

```protobuf
repeated VolumeCapability volume_capabilities = 3;
```

I couldn't remember why we did that? When creating a volume why would we want to specify more then one capability? AFAIK both `access_type` and `access_mode` can be specified in a `VolumeCapability` object.

I vaguely recall a discussion about not knowing exactly how a volume will be consumed at provision time, but I can't recall if this would help that case.

So purpose of this bug is to answer the question: why is `VolumeCapability` a list in `CreateVolumeRequest`?

**Comment by jieyu** [comment]
@saad-ali IIRC, the intention was to allow SP to provision a volume that can be published as both Mount or Block volume. Think about creating an EBS volume that can be used as both a Block volume or a Mount volume. At the creation time, the CO might not know the exact way that the volume will be published.

**Comment by jdef** [comment]
If we don't already, we should probably call out that scenario in the spec,
for clarity

**Comment by msau42** [comment]
If the capability passed in is Mount, is it expected for a block volume plugin to also create the filesystem?  Or is the CO expected to handle filesystem formatting?

**Comment by jieyu** [comment]
@msau42 SP is eventually responsible for formatting the filesystem. When does the SP decide to format the filesystem is up to the SP. I imagine most SP will do that in NodePublishVolume, but that's not a requirement in the spec.

**Comment by msau42** [comment]
I guess I'm kind of confused as a plugin writer about what I should do when given multiple VolumeCapabilities.  Do I pick one of them to support? Or do I need to provision a volume in such a way to support all of them?  There is no option in the VolumeCreateResponse to specify which VolumeCapability I chose.

**Comment by msau42** [comment]
Another aspect I'm a little confused about are for volume plugins that can only support file or block (but not both).  How do I report to the CO that capability?

**Comment by jieyu** [comment]
@msau42 I think the wording in the spec is pretty clear. If the SP cannot support ANY of the capability requested by the CO, an error should be returned.

```// The capabilities that the provisioned volume MUST have: the Plugin
  // MUST provision a volume that could satisfy ALL of the
  // capabilities specified in this list. The Plugin MUST assume that
  // the CO MAY use the  provisioned volume later with ANY of the
  // capabilities specified in this list. This also enables the CO to do
  // early validation: if ANY of the specified volume capabilities are
  // not supported by the Plugin, the call SHALL fail. This field is
  // REQUIRED.
  repeated VolumeCapability volume_capabilities = 3;```

**Comment by msau42** [comment]
So for volume plugins that can only support block, or only support mount, how does it tell the CO what it supports?  I think there is some capability discovery mechanism missing.

**Comment by jieyu** [comment]
I think capability discovery is not just for Mount vs. Block. For instance, how does CO know what access mode a plugin supports, what mount options a plugin supports?

I remembered that we had a debate on this, the consensus is that it's very hard to enumerate all possible capabilities that a plugin supports. Instead, let CO specify what capabilities it needs based on user request.

**Comment by msau42** [comment]
So if the CO passes in capabilities based on the user's request, then shouldn't it know exactly what capability it needs?

For example, NFS only supports Mount access type.  If the CO doesn't know exactly what access type to specify, and it passes in both Block and Mount, then the NFS plugin will always fail the CreateVolume request, and it has no way to tell the CO to stop passing in the Block capability.

**Comment by jieyu** [comment]
I think we should separate the discussion of plugin capability discovery from discussion. IMO, this is a separate issue. Even if say we make `VolumeCapability` not a list in `CreateVolumeRequest`, we will have the same issue. If the CO passes in a mount option that the SP does not support, the plugin will always fail the `CreateVolumeRequest`, and has no way to tell the CO to stop passing the unsupported mount option.

I think the original purpose for making `VolumeCapability` a list in `CreateVolumeRequest` is to allow a CO to provision a volume that can be later used as both Block or Mount volume. I think it's also for the case to provision a volume supports multiple filesystems/access modes. This allows some CO to *delay* the decision of how to use the volume when provision a volume.

But I am unaware any CO leverage this feature yet.

**Comment by msau42** [comment]
I think this implies then that the SP cannot apply any specific access type or access mode at CreateVolume time, and only at NodePublishVolume time, since the volume created needs to be able to support all the capabilities passed in.

**Comment by jieyu** [comment]
@msau42 Yeah, that's true. I think that was the intention. Would love to hear if there's any concrete problem you saw with this?

**Comment by gnufied** [comment]
This was brought up in slack. When a volume supports more than one access mode - for example `MULTI_NODE_MULTI_WRITER` does that automatically implies `SINGLE_NODE_WRITER` or does CO should specify both the capabilities as individual items?

**Comment by saad-ali** [comment]
I was thinking through this with @thockin and couldn't defend why it was a list. Let's make sure we finalize on this before 1.0.

**Comment by jdef** [comment]
should align #271 with our decision here

**Comment by jdef** [comment]
@jieyu and I discussed this and I think we're OK making this a non-repeated field. I'm happy to address this as part of #271 as well (since that PR already contains breaking changes).

**Comment by jdef** [comment]
Or, since it may be a more involved change, we can land it after we come to agreement on #271?

**Comment by jieyu** [comment]
@saad-ali do you want to send a PR. We can take it if you don't have time. It'll be nice to land this before 1.0 because it's a breaking change.

**Comment by saad-ali** [comment]
Thought about this more. Use case that k8s has for this is k8s allows user to request multiple access modes and plugin should be able to get that info and reject if it can't support it.

I'll add this use case to the comment to clarify.
